picard/util/alphabet_detector.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+# The MIT License (MIT)
+#
+# Copyright (C) 2015 Eli Finkelshteyn
+# Copyright (C) 2021 Vladislav Karbovskii
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from collections import defaultdict
+import unicodedata as ud
+
+
+class AlphabetDetector:
+    def __init__(self, no_memory=False):
+        self.alphabet_letters = defaultdict(dict)
+        self.no_memory = no_memory
+
+    def is_in_alphabet(self, uchr, alphabet):
+        if self.no_memory:
+            return alphabet in ud.name(uchr)
+        try:
+            return self.alphabet_letters[alphabet][uchr]
+        except KeyError:
+            return self.alphabet_letters[alphabet].setdefault(
+                uchr, alphabet in ud.name(uchr))
+
+    def only_alphabet_chars(self, unistr, alphabet):
+        return all(self.is_in_alphabet(uchr, alphabet)
+                   for uchr in unistr if uchr.isalpha())
+
+    def detect_alphabet(self, unistr):
+        return set(ud.name(char).split(' ')[0]
+                   for char in unistr if char.isalpha())
```

Comment:
To be honest I needed working solution for this problem and it exists (https://github.com/EliFinkelshteyn/alphabet-detector). But instead of having additional dependency I decided to assimilate it with slight changes. It is simple, it works, it was tested by main author, and license allows to do it. If you say we have to rewrite it - we could but should we?

Reply:
Anyway done with simplification, will add test later

Reply:
Test for alphabet detection was added